Source/Dafny/AST/DafnyAst.cs

```diff
@@ -9070,6 +9086,48 @@ public abstract class Expression {
       get { yield break; }
     }
 
+    private IToken startTok;
+    private IToken endTok;
+
+    // Creates a token on the entire range of the expression.
+    // Used only for error reporting.
+    public IToken RangeToken => new RangeToken(StartTok, EndTok);
```

Comment:
Do you think we could move all the new code to a class 

```
class SyntaxNode {
  public readonly IToken tok;

  public abstract IEnumerable<SyntaxNode> ChildNodes {get;}

  public IToken RangeToken =>

  ..
}

``` 

and then have

```
class Expression : SyntaxNode {
  public IEnumerable<SyntaxNode> ChildNodes => SubExpressions;
}

```

Or do you feel that's too early for this PR ?

Reply:
We could as well use the bounded polymorphism pattern:

```
class SyntaxNode<T> where T: SyntaxNode {
  public readonly IToken tok;

  public abstract IEnumerable<T> Children{get;}

  public IToken RangeToken =>
  ..
}
class Expression : SyntaxNode<Expression> {
}
```

However, we should do this refactoring later, when I see a compelling reason to do so. We might as well replace the "tok" by a Format or trait to keep track of concrete syntax tree elements.